### PR TITLE
Add ttnn.cumprod scaffold

### LINKED ISSUE #19549: Add ttnn.cumprod scaffold
Following cumulative product definition that can be found here, implement ttnn.cumprod scaffold.
https://pytorch.org/docs/stable/generated/torch.cumprod.html

Add ttnn.cumprod to reduction ops group that can be found here:
tt-metal/ttnn/cpp/ttnn/operations/experimental/reduction/

Add a few unit tests that iterate through different input parameter variations to directory:
tt-metal/tests/ttnn/unit_tests/operations/

The scaffold for the operation should just return the correct tensor shape based on input parameters, it is not required yet that the tensor contain the right values inside it.

### PR DESCRIPTION
**Pull request by jbbieniekTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/19549

### Problem description
Implement scaffold in terms of `ttnn.cumprod` device operation. It returns the same tensor it was supplied with the same size and dtype. The full implementation will be provided later. 

### What's changed
- The approach used to add `ttnn.cumprod` is derived from https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/adding_new_ttnn_operation.html.
- There is a `test_cumprod.py` Pytest UT added.

(links to pipelines incoming)
### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [X] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [X] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [X] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/test_cumprod.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import pytest
+
+import torch
+
+import ttnn
+
+
+@pytest.mark.parametrize("dim,", [0, 1, 2, 3])
+@pytest.mark.parametrize("shape", [[1], [1, 2], [1, 2, 3], [1, 2, 3, 4]])
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/test_cumprod.py:13]: These are all sub-tile shapes. It's good to have a combination of

- sub tile (e.g. 3 or 16),
- tile (32),
- multi tile (e.g. 64 or 128), and
- multi tile offset (e.g. 87).
Also, be careful about how large the tensors get, which would affect how long the tests take to run.

File: tests/ttnn/unit_tests/operations/test_cumprod.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+
+import torch
+
+import ttnn
+
+
+@pytest.mark.parametrize("dim,", [0, 1, 2, 3])
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/test_cumprod.py:12]: The cross product of the parameters will be taken. Not all of these dims will work with all of the shapes.

If you want to go through all possible dims, have a for loop in the test.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.cpp [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "device/cumprod_device_operation.hpp"
+#include "cumprod.hpp"
+
+namespace ttnn::operations::experimental::reduction {
+
+// TODO(jbbieniek): add doc
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.cpp:10]: What do you mean by add doc?

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "ttnn/decorators.hpp"
+
+namespace ttnn {
+namespace operations::experimental::reduction {
+
+struct CumprodOperation {
+    static Tensor invoke(
+        const Tensor& input_tensor,
+        int64_t dim  // TODO(jbbieniek): int8_t?
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:15]: Just use int32_t throughout for dim, including pybind.

**Review by bbradelTT — COMMENTED** [review]
Although the single and multi core looks like copied from the existing example, having two approaches means double the work and half the velocity.

Unless you have a strong reason to have two approaches, you should just have one.

Just have one approach that calculates how many cores to use, from 1 to n.

Also, there may already be existing common code to split the work across cores.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+        Returns a tensor witth cumulative product calculated along a given axis (`dim`).
+
+        Args:
+            input_tensor (ttnn.Tensor): the input tensor to calculate cumulative product of.
+            dim (int): direction of product cumulation
+
+        Returns:
+            ttnn.Tensor: the output tensor (for now, it's same as input_tensor, because only scaffold is implemented).
+        )doc";
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:32]: Please add a short code snippet example.

Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:32]: Coming in the next commit.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.cpp (hunk 3 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.cpp:10]: Forgot to remove this. I will also have a look at all my TODOs.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 5 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:38]: I have committed the change.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp (hunk 4 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp]: I have committed the change.

File: tests/ttnn/unit_tests/operations/test_cumprod.py (hunk 1 above)
Comment (jbbieniekTT) [reply tests/ttnn/unit_tests/operations/test_cumprod.py]: I have committed the change. Please check if this is what we're looking for.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: Please split into hpp and cpp.
hpp should only include fwd declaration of py module.

**Review by bbradelTT — APPROVED** [review]
Just one minor comment.

File: tests/ttnn/unit_tests/operations/test_cumprod.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        [1, 2, 3, 4],
+        [3, 3],
+        [16, 16],
+        [32, 32, 32],
+        [32, 32, 64],
+        [64, 64, 64, 64],
+        [1000, 32, 32],
+    ],
+)
+@pytest.mark.parametrize("dtype", [(torch.bfloat16, ttnn.bfloat16)])
```
Comment (philei-tt) [inline tests/ttnn/unit_tests/operations/test_cumprod.py:28]: What's the point of adding a parameter with one possible value?

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace ttnn::operations::experimental::reduction {
+
+using tt::tt_metal::MemoryConfig;
+using tt::tt_metal::Tensor;
+using tt::tt_metal::TensorLayout;
+using tt::tt_metal::TensorSpec;
+
+struct CumprodDeviceOperation {
+    struct operation_attributes_t {
```
[the remark this answers is not in the document]
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.hpp:24]: Okay

File: tests/ttnn/unit_tests/operations/test_cumprod.py (hunk 7 above)
Comment (jbbieniekTT) [reply tests/ttnn/unit_tests/operations/test_cumprod.py:28]: I am going to add more dtypes once the implementation of `ttnn.cumprod` is complete. For now, I left room for parametrising more types.

File: tests/ttnn/unit_tests/operations/test_cumprod.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        [1, 2],
+        [1, 2, 3],
+        [1, 2, 3, 4],
+        [3, 3],
+        [16, 16],
+        [32, 32, 32],
+        [32, 32, 64],
+        [64, 64, 64, 64],
+        [1000, 32, 32],
+    ],
```
[the remark this answers is not in the document]
Comment (jbbieniekTT) [reply tests/ttnn/unit_tests/operations/test_cumprod.py:26]: Done

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 6 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: Done

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.cpp [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#include "cumprod_device_operation.hpp"
+
+namespace ttnn::operations::experimental::reduction {
+
+// the result depends on tensor_args!
+CumprodDeviceOperation::program_factory_t CumprodDeviceOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return CumprodProgramFactory{};  // TODO(jbbieniek): enable multi-core version once the implementation is ready
```
[the remark this answers is not in the document]
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.cpp:12]: Removed

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 6 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: Done

Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: Okay

File: tests/ttnn/unit_tests/operations/test_cumprod.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (philei-tt) [reply tests/ttnn/unit_tests/operations/test_cumprod.py]: 0 rank is []

[the remark this answers is not in the document]
Comment (bbradelTT) [reply tests/ttnn/unit_tests/operations/test_cumprod.py]: Please keep the [0]. That's a 0 volume tensor, and it would be good to test too.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 6 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: I have managed to do this now :)

File: tests/ttnn/unit_tests/operations/test_cumprod.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbbieniekTT) [reply tests/ttnn/unit_tests/operations/test_cumprod.py]: Done

**Comment by ayerofieiev-tt** [comment]
@jbbieniekTT please update the PR title as it will become the commit message in main after merge.

**Review by sminakov-tt — APPROVED** [review]

**Review by philei-tt — APPROVED** [review]

**Comment by jbbieniekTT** [comment]
@razorback3 @TT-BrianLiu @cfjchu @dongjin-na @ayerofieiev-tt @dmakoviichuk-tt @rfurko-tt @omilyutin-tt
Hello, just a gentle reminder to review my changes. :)

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 6 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: Please remove all includes from here, only use fwd include from pybind folder.

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <optional>
+#include "ttnn/decorators.hpp"
+
+namespace ttnn {
+namespace operations::experimental::reduction {
+
+struct CumprodOperation {
+    static Tensor invoke(
+        const Tensor& input_tensor,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:15]: please add `QueueId cq_id ` argument here. 
it allows to dispatch a given operation to a specific command queue

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp (hunk 6 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod_pybind.hpp:5]: https://github.com/tenstorrent/tt-metal/blob/5622c69f846aa6982ab27571214dc784cde93fd0/ttnn/cpp/ttnn/operations/experimental/dropout/dropout_pybind.hpp#L6

**Review by ayerofieiev-tt — APPROVED** [review]
You might have to add a change to rst similar to ongoing ttnn.sort PR

File: ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp (hunk 11 above)
Comment (jbbieniekTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:15]: Done
